trin-core/src/utp/packets.rs

```diff
@@ -215,67 +193,87 @@ impl AsRef<[u8]> for Packet {
     }
 }
 
+impl Default for Packet {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
 impl Packet {
-    pub fn new(mut packet_header: PacketHeader) -> Packet {
-        let mut vec = Vec::with_capacity(HEADER_SIZE);
-        vec.append(&mut packet_header.encode());
-        Packet(vec)
+    pub fn new() -> Packet {
+        Packet(PacketHeader::default().as_ref().to_owned())
```

Comment:
This is a surprising pattern that maybe I just haven't run into before. Why build a new object, take a reference, and then convert it back to owned (presumably incurring the cost of copying along the way)?

Reply:
We have `AsRef` trait implemented for `PacketHeader` here:

```rust
impl AsRef<[u8]> for PacketHeader {
    /// Returns the packet header as a slice of bytes.
    fn as_ref(&self) -> &[u8] {
        unsafe { &*(self as *const PacketHeader as *const [u8; HEADER_SIZE]) }
    }
}
```

which is returning a slice of bytes and one way to convert from a slice of bytes to `Vec<u8>` is `to_owned()`. It is similar to converting from `&str` to `String`.